distributed/worker.py

```diff
@@ -2139,6 +2145,13 @@ def transition_executing_released(self, ts, *, stimulus_id):
         # See https://github.com/dask/distributed/pull/5046#discussion_r685093940
         ts.state = "cancelled"
         ts.done = False
+
+        if self.interruptor:
+            th = [th for th, k in self.active_threads.items() if k == ts.key]
+            if th:
+                logger.info("Interrupting thread %i for task %s", th[0], ts.key)
+                self.executor.interrupt(th[0])
```

Comment:
I would argue it is not possible for `th` to be empty. If that is true, there should probably be a `else: RuntimeError`. Although I'm a bit ambivalent since on the one hand I'd like us to avoid implicit fail cases but on the other hand, we are not raising exceptions anywhere in the transition engine, so far.

Reply:
I imagine it's possible for the task to end at the same time as it is released for another reason. If it turns out there is no longer a thread associated, that doesn't seem like an error to me.

Reply:
Good point. Can we test this?

Reply:
At the very least, we should add a comment here. It's not straight forward to figure out and it feels like an important detail.

I'm also wondering now what happens if the thread finishes properly while we're trying to interrupt vs the task raised an exception before and we're trying to interrupt. What will be the final task state if either one happens? (I'm mostly concerned about invalid states. If we rerun anything, I don't mind as long as the state machine isn't messed up)

Reply:
> Can we test this?

I'm don't see a way we can emulate it without horrible race problems.

> what happens if the thread finishes properly while we're trying to interrupt

The thread doesn't finish as such, it goes into the loop to pick more work, which is what [this](https://github.com/dask/distributed/pull/4726/files#diff-894cd176005fcf92cb8b6a46d6a4efdb3644d23ef81928ad2ec91c1190e671d9R114) is for. Having said that, the _WorkItem s don't have a state beyond what's encoded in the corresponding future.

> task raised an exception before and we're trying to interrupt

If the task is already in an Except block, you would expect to see "During handling of the above exception, another exception occurred:" as the return, but `finally:` blocks would still run.

> If we rerun anything, I don't mind as long as the state machine isn't messed up

Fundamentally, we are cleaning up a task which is set for release, and so the output of the task and the Item corresponding object's state don't matter.